sharding/simulator/service.go

```diff
@@ -13,82 +13,67 @@ import (
 	"github.com/ethereum/go-ethereum/sharding/p2p/messages"
 	"github.com/ethereum/go-ethereum/sharding/params"
 	"github.com/ethereum/go-ethereum/sharding/syncer"
+	"github.com/ethereum/go-ethereum/sharding/utils"
 )
 
-// Simulator is a service in a shard node
-// that simulates requests from remote notes coming over
-// the shardp2p network. For example, if we are running a
-// proposer service, we would want to simulate notary requests
-// coming to us via a p2p feed. This service will be removed once
-// p2p internals and end-to-end testing across remote nodes have been
-// implemented.
+// Simulator is a service in a shard node that simulates requests from
+// remote notes coming over the shardp2p network. For example, if
+// we are running a proposer service, we would want to simulate notary requests
+// requests coming to us via a p2p feed. This service will be removed
+// once p2p internals and end-to-end testing across remote
+// nodes have been implemented.
 type Simulator struct {
 	config      *params.Config
-	reader      mainchain.Reader
-	fetcher     mainchain.RecordFetcher
+	client      *mainchain.SMCClient
 	p2p         *p2p.Server
 	shardID     int
 	ctx         context.Context
 	cancel      context.CancelFunc
-	errChan     chan error    // Useful channel for handling errors at the service layer.
-	delay       time.Duration // The delay (in seconds) between simulator requests sent via p2p.
-	requestSent chan int      // Useful channel for handling outgoing requests from the service.
+	errChan     chan error // Useful channel for handling errors at the service layer.
+	delay       time.Duration
+	requestFeed *event.Feed
 }
 
 // NewSimulator creates a struct instance of a simulator service.
 // It will have access to config, a mainchain client, a p2p server,
 // and a shardID.
-func NewSimulator(config *params.Config, reader mainchain.Reader, fetcher mainchain.RecordFetcher, p2p *p2p.Server, shardID int, delay time.Duration) (*Simulator, error) {
+func NewSimulator(config *params.Config, client *mainchain.SMCClient, p2p *p2p.Server, shardID int, delay time.Duration) (*Simulator, error) {
```

Comment:
I thought we wanted to use `reader` and `fetcher` for smaller interfaces to make it easier to test. What's the benefit here of switching back to `SMCClient`?

Reply:
I think the reason was that reader and fetcher refer to the same object. So it didnt make sense to add two arguments instead of one. Although it makes harder to test now

Reply:
@nisdas is correct that they refer to the same object. Also it does not make it hard to test, as we can just set properties directly into the simulator during the testing process (see my test files for example).

Another big reason we cannot do this is that reader and fetcher are not available during the service registration process, only after Start(), so we have to use mainchain.SMCClient as an argument.